bug: e2e: cgroupfs mount failure in OCI mode with runc

**Version of Singularity**
What version of Singularity are you using? Run:

**Describe the bug**

Seen with Ubuntu 22.04 and runc.

```
--- FAIL: TestE2E/SEQ/CGROUPS/action_flags_rootless_cgroups/rootless/OCIUser (0.78s)
                            --- FAIL: TestE2E/SEQ/CGROUPS/action_flags_rootless_cgroups/rootless/OCIUser/memory (0.22s)
                            --- FAIL: TestE2E/SEQ/CGROUPS/action_flags_rootless_cgroups/rootless/OCIUser/memory-reservation (0.18s)
                            --- FAIL: TestE2E/SEQ/CGROUPS/action_flags_rootless_cgroups/rootless/OCIUser/memory-swap (0.18s)
                            --- FAIL: TestE2E/SEQ/CGROUPS/action_flags_rootless_cgroups/rootless/OCIUser/pids-limit (0.20s)
```

Underlying cause is that the tests attempt to bind the cgroupfs into the container to observe the limits being applied.

This fails with runc (succeeds with crun):

```
    === RUN   TestE2E/SEQ/CGROUPS/action_flags_rootless_cgroups/rootless/OCIUser/memory
        singularitycmd.go:698: Running command "/usr/bin/singularity exec --oci --memory 500M -B /sys/fs/cgroup oci-sif:/tmp/stest.2268700094/oci-sif.sif /bin/sh -c cat /sys/fs/cgroup$(cat /proc/self/cgroup | grep '^0::' | cut -d ':' -f 3)/memory.max"
        singularitycmd.go:698:
            "/usr/bin/singularity exec --oci --memory 500M -B /sys/fs/cgroup oci-sif:/tmp/stest.2268700094/oci-sif.sif /bin/sh -c cat /sys/fs/cgroup$(cat /proc/self/cgroup | grep '^0::' | cut -d ':' -f 3)/memory.max" output:
            time="2023-08-16T17:09:39Z" level=error msg="runc run failed: unable to start container process: error during container init: error mounting \"/sys/fs/cgroup\" to rootfs at \"/sys/fs/cgroup\": mount /proc/self/fd/7:/sys/fs/cgroup (via /proc/self/fd/11), flags: 0x5026: operation not permitted"
  
        singularitycmd.go:698: got 1 as exit code and was expecting 0: exit status 1
            waiting for command "/usr/bin/singularity exec --oci --memory 500M -B /sys/fs/cgroup oci-sif:/tmp/stest.2268700094/oci-sif.sif /bin/sh -c cat /sys/fs/cgroup$(cat /proc/self/cgroup | grep '^0::' | cut -d ':' -f 3)/memory.max"
            github.com/sylabs/singularity/v4/e2e/internal/e2e.TestEnv.RunSingularity.func1
                    github.com/sylabs/singularity/v4/e2e/internal/e2e/singularitycmd.go:673
            github.com/sylabs/singularity/v4/e2e/internal/e2e.TestEnv.RunSingularity.func2
                    github.com/sylabs/singularity/v4/e2e/internal/e2e/singularitycmd.go:698
            runtime.goexit
                    runtime/asm_arm64.s:1197
```

--- a/internal/pkg/runtime/launcher/oci/mounts_linux.go
+++ b/internal/pkg/runtime/launcher/oci/mounts_linux.go
@@ -625,12 +625,18 @@ func (l *Launcher) addBindMount(mounts *[]specs.Mount, b bind.Path, allowSUID bo
 		)
 	}
 
+	// /proc and /sys need nosuid,noexec,nodev set explicitly for runc
+	// See: https://github.com/opencontainers/runc/discussions/3801
 	opts := []string{"rbind", "nodev"}
+	if strings.HasPrefix(b.Source, "/proc") || strings.HasPrefix(b.Source, "/sys") {
+		opts = append(opts, "nosuid", "noexec")
+	} else {
+		if !allowSUID {
+			opts = append(opts, "nosuid")
+		}
+	}
 	if b.Readonly() {
 		opts = append(opts, "ro")
-	}
-	if !allowSUID {
-		opts = append(opts, "nosuid")
 	}
 
 	absSource, err := filepath.Abs(b.Source)

--- a/internal/pkg/runtime/launcher/oci/mounts_linux_test.go
+++ b/internal/pkg/runtime/launcher/oci/mounts_linux_test.go
@@ -57,7 +57,7 @@ func Test_addBindMount(t *testing.T) {
 					Source:      "/tmp",
 					Destination: "/tmp",
 					Type:        "none",
-					Options:     []string{"rbind", "nodev", "ro", "nosuid"},
+					Options:     []string{"rbind", "nodev", "nosuid", "ro"},
 				},
 			},
 		},
@@ -119,6 +119,38 @@ func Test_addBindMount(t *testing.T) {
 			wantMounts: &[]specs.Mount{},
 			// Should fail because bind-mounting SIFs not supported in OCI mode
 			wantErr: true,
+		},
+		{
+			name: "Proc",
+			b: bind.Path{
+				Source:      "/proc",
+				Destination: "/proc",
+			},
+			userbind: true,
+			wantMounts: &[]specs.Mount{
+				{
+					Source:      "/proc",
+					Destination: "/proc",
+					Type:        "none",
+					Options:     []string{"rbind", "nodev", "nosuid", "noexec"},
+				},
+			},
+		},
+		{
+			name: "Sys",
+			b: bind.Path{
+				Source:      "/sys",
+				Destination: "/sys",
+			},
+			userbind: true,
+			wantMounts: &[]specs.Mount{
+				{
+					Source:      "/sys",
+					Destination: "/sys",
+					Type:        "none",
+					Options:     []string{"rbind", "nodev", "nosuid", "noexec"},
+				},
+			},
 		},
 	}
 	for _, tt := range tests {
@@ -202,7 +234,7 @@ func TestLauncher_addBindMounts(t *testing.T) {
 					Source:      "/tmp",
 					Destination: "/mnt",
 					Type:        "none",
-					Options:     []string{"rbind", "nodev", "ro", "nosuid"},
+					Options:     []string{"rbind", "nodev", "nosuid", "ro"},
 				},
 			},
 			wantErr: false,
@@ -287,7 +319,7 @@ func TestLauncher_addBindMounts(t *testing.T) {
 					Source:      "/tmp",
 					Destination: "/mnt",
 					Type:        "none",
-					Options:     []string{"rbind", "nodev", "ro", "nosuid"},
+					Options:     []string{"rbind", "nodev", "nosuid", "ro"},
 				},
 			},
 			wantErr: false,
@@ -408,7 +440,7 @@ func TestLauncher_addLibrariesMounts(t *testing.T) {
 					Source:      lib1,
 					Destination: "/.singularity.d/libs/lib1.so",
 					Type:        "none",
-					Options:     []string{"rbind", "nodev", "ro", "nosuid"},
+					Options:     []string{"rbind", "nodev", "nosuid", "ro"},
 				},
 			},
 			wantErr: false,
@@ -424,13 +456,13 @@ func TestLauncher_addLibrariesMounts(t *testing.T) {
 					Source:      lib1,
 					Destination: "/.singularity.d/libs/lib1.so",
 					Type:        "none",
-					Options:     []string{"rbind", "nodev", "ro", "nosuid"},
+					Options:     []string{"rbind", "nodev", "nosuid", "ro"},
 				},
 				{
 					Source:      lib2,
 					Destination: "/.singularity.d/libs/lib2.so",
 					Type:        "none",
-					Options:     []string{"rbind", "nodev", "ro", "nosuid"},
+					Options:     []string{"rbind", "nodev", "nosuid", "ro"},
 				},
 			},
 			wantErr: false,

